---
title: "Common pitfalls in estimating contamination"
author: "Matthew Daniel Young"
date: "`r Sys.Date()`"
fig_width: 8
fig_height: 6
output:
  pdf_document: default
  html_document: default
vignette: >
  %\VignetteIndexEntry{Common pitfalls in estimating contamination}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r global_options, include=FALSE}
library(knitr)
opts_chunk$set(tidy=TRUE)
```

# Introduction

The most complicated and error prone part of the decontamination method implemented in [SoupX](https://github.com/constantAmateur/SoupX) is the estimation of contamination fraction.  This vignette will cover some of the most common mistakes made in estimating the fraction of ambient RNA present in each cell and review the best practice.

To quantify the amount of contamination present in each cell, the aim is to identify a set of genes that are not expressed in most cells, assume that any observed expression of these genes must derive from the soup and then use the observed expression to infer the contamination fraction.  Because the expression of a set of genes is usually low, it is necessary to bin together cells (done by total UMIs per cell by default) to obtain a sufficiently accurate estimate.

The key points to note are that:

 - It is necessary to specify a set of genes that are not expressed.
 - It is necessary to specify which cells do not express this set of genes.
 - Each cell's contamination fraction usually needs to be interpolated from binned estimates.

Most of the time when contamination estimates go wrong it is because one of these points has been misunderstood.  When in doubt, the over-riding principle you should follow is, "get the lowest estimate that still makes sense".

# Best practice demonstration using PBMC

The ideal way to approach the estimate of the contamination is to maximise the use of prior biological knowledge.  In practice this means annotating each cell and assigning it to a cell type, then deciding which genes should not be expressed for each cell type.  Taking the PBMC data as an example, we start by processing, clustering and annotating the uncorrected data.  This produces the following map,

![Annotated PBMC data](../images/PBMC_Annotation.png)

We know that immunoglobulin genes should not be expressed by NK or T-Cells and so we would like to use this information to estimate the contamination.  To do this we first load the soup channel list object created from the PBMC data in the other vignette and the annotation data

```{r load_scl}
library(SoupX)
data(PBMC_DR)
data(PBMC_SCL)
```

Clusters 1,2,3,5,6 and 8 all represent T or NK cells.  We extract the unique labels of the cells in these clusters to use,

```{r extractCellLabs}
labs = rownames(PBMC_DR)[PBMC_DR$Cluster %in% c(1,2,3,5,6,8)]
```

We want to tell `calculateContaminationFraction` to that we know that immunoglobulin genes should not be expressed in T or NK cells and so it can use the presence of these genes to estimate the contamination rate.  To do this we construct a matrix saying which cells to use igGenes in

```{r useToEst}
toUse = matrix(colnames(PBMC_SCL$toc) %in% labs,nrow=1,dimnames=list('IG',colnames(PBMC_SCL$toc)))
```

Now we define the IG genes and estimate the contamination

```{r estRho}
igGenes = c('IGHA1','IGHA2','IGHG1','IGHG2','IGHG3','IGHG4','IGHD','IGHE','IGHM',
            'IGLC1','IGLC2','IGLC3','IGLC4','IGLC5','IGLC6','IGLC7',
            'IGKC')
PBMC_SCL = calculateContaminationFraction(PBMC_SCL,'Channel1',list(IG=igGenes),useToEst=toUse)
```

Let's plot the resulting contamination estimates,

```{r estRhoPlot1}
plotChannelContamination(PBMC_SCL,'Channel1')
```

And compare this with what we would have got if we'd used the default values as in the PBMC vignette.

```{r estRhoPlot2}
PBMC_SCL = calculateContaminationFraction(PBMC_SCL,'Channel1',list(IG=igGenes))
plotChannelContamination(PBMC_SCL,'Channel1')
```

Why is there a huge spike towards the high UMI end of the plot not present in the other data?  To understand this, we need to understand what `calculateContaminationFraction` does when not supplied with the `useToEst` argument.  In this case the code has to guess which cells truly express each set of genes (in our case, just the IG genes) and which express it as a consequence of the contamination.  The only information it has to go on is the expression level of the gene set in each cell.  It therefore uses this information to guess at which cells to use to estimate the contamination.

By default, this is done using a Poisson test.  The null hypothesis is that each cell is composed entirely of soup.  Under this null hypothesis, the code then calculates how many IG genes would be expected in this cell based on the total number of UMIs.  The alternative hypothesis is that the droplet contains a cell that highly expresses IG genes.  If this is true, then IG gene expression will be greater than what we would expect under the null as the cell will be actively creating extra IG gene transcripts.  `calculateContaminationFraction` then excludes any gene for which the p-value (after BH multiple hypothesis correction) is less than `exCut` (0.05 by default).

We can see which cells this test excludes by running the estimation function ourselves

```{r useToEstAuto}
toUsePoisson = identifyExpressingCells(PBMC_SCL,'Channel1',list(IG=igGenes))
toUsePoisson[,1:10]
table(toUsePoisson[1,])
```

We see that this test has excluded cells mostly from clusters 4 and 7, which are the B-cell clusters in our above annotation.  We also note that the test has concluded that there are 2 cells that we have marked as "NK" but appear to be genuinely expressing IG genes.  The test usually does a good job of excluding the cells that are really clearly expressing a set of genes, but cannot exclude those that express the gene but at a moderate level.

Let's recalculate the contamination but excluding these two cells.

```{r useToEstComb}
toUseComb = toUse & toUsePoisson
sum(toUse)
sum(toUseComb)
PBMC_SCL = calculateContaminationFraction(PBMC_SCL,'Channel1',list(IG=igGenes),useToEst = toUseComb)
plotChannelContamination(PBMC_SCL,'Channel1')
```

The resulting contamination plot is almost the same, but no longer has the spike going up to 12% contamination.

## Interpolation

The estimation of the contamination fraction produced is not per cell, but per a group of cells.  This is done because each individual cell often does not have enough molecules that we can be certain are contamination (IG genes in T/NK cells in our example) to get an accurate estimate of the contamination fraction.  By default, `calculateContaminationFraction` will group together enough cells such that there would be roughly 1000 counts for IG genes (or whatever the known contamination set are) if those cells were nothing but soup.  This errs on the side of grouping cells together as the change in contamination rate with number of UMIs is rather gradual in most instances.  We can decrease the number of cells in each bin by reducing `tgtSoupCntsPerGroup` or explicitly indicating which cells should be grouped together by providing the `cellGroups` argument.

```{r smallerBins}
PBMC_SCL = calculateContaminationFraction(PBMC_SCL,'Channel1',list(IG=igGenes),useToEst = toUseComb,tgtSoupCntsPerGroup=100)
plotChannelContamination(PBMC_SCL,'Channel1')
```

Which reveals that there are probably still two cells being used to estimate the contamination which we shouldn't be using.  We could try and identify these cells and remove them, but a more useful strategy is to instead rely on the majority of bins being accurate and smoothing the estimates with a lowess curve.  

This smoothed estimate has actually been present in every plot we've made so far.  Each contamination plot contains a red line showing the global average and the confidence interval around it, a green line showing a lowess smoothed estimate and a blue line showing a lowess estimate constrained to monotonically decrease as a function of total UMI count.  By default, it is this final curve, the monotonically decreasing lowess curve, that is used to assign each cell a contamination value.  The curve is constrained to monotonically decrease as given a constant number of contaminating molecules per-droplet, the contamination fraction will decrease as the number of cell derived mRNA molecules increases (i.e., as the total number of UMIs increases).

To assign each cell its own contamination estimate, we simply run

```{r interpolate}
PBMC_SCL = interpolateCellContamination(PBMC_SCL,'Channel1')
```

which will approximate each cell's contamination value using the monotonically decreasing lowess curve.  It can often be useful to manually specify a global contamination value, particularly for channels such as this where the contamination rate is fairly constant as a function of total number of UMIs.  For example, the following command will set every cell to have a contamination fraction of 5%

```{r fixedRho}
PBMC_SCL = interpolateCellContamination(PBMC_SCL,'Channel1',interpolationMethod='fixed',fixedContaminationValue=0.05)
```

# Which genes should I use?

Most of the time, the user should have sufficient biological knowledge to know that there is a certain set of genes that should not be present in most cell types (e.g. Immunoglobulin genes and B-cells, Haemoglobin genes and red blood cells).  However, sometimes it will not be obvious a-priori which set of genes to use, or the set of genes you had hoped to use may be completely absent from your experiment, rendering it uninformative.

Generally, the sets of genes that are useful are those that are very highly expressed when they are expressed, very specific to one cell type and the cell type that expresses them is present in your data.  This is because when a set of genes either highly expressed or not expressed at all, it is easy to separate the cells that truly express these genes from those that only appear to express it due to contamination.  The cell type needs to be present in order for the soup to contain the target set of genes.

Thinking about these properties, the genes that are most useful should have a biomodal expression distribution across all cells, with a low level of expression present in those cells that do not express it but look like they do due to contamination and a high level of expression from those cells that genuinely express the gene.  The purpose of the `inferNonExpressedGenes` function is to provide the user with a list of potential candidate genes that *may or may not* be suitable for estimating the contamination rate.

Applying this function to the PBMC data,

```{r inferGenes}
PBMC_SCL = inferNonExpressedGenes(PBMC_SCL)
plotMarkerDistribution(PBMC_SCL,'Channel1')
```

This plot is rather complicated and needs some explanation.  For each gene shown, we look in every cell and calculate two things:

 1. The expected number of counts of gene g if cell c were pure soup.  That is, if cell c has $latex N_{UMI}$ total UMIs and gene g has normalised expression $latex f_{gs} = \frac{n_{gs}}{\sum_g {n_{gs}}}$ where $latex n_{gs}$ is the number of counts of gene g in the empty droplets, then the expected number of counts is calculated as $latex f_{gs} N_{UMI}$.
 2. The observed number of counts of gene g in cell c.

For each gene we then plot the distribution of the log ratio of observed to expected counts (the violin plot) and show this value for a randomly sampled set of 100 cells (circles).  The same hypothesis test that is used by `calculateContaminationFraction` is used to determine which of these cells definitively express each gene and these are marked in red.  Finally, the red circles and line indicate what the average global contamination would be estimated to be if that gene alone were used to estimate the contamination.

The ideal gene has two modes to its distribution, one well above zero, the other below, which produces a good separation of red and black dots.  This plot shows the top 20 genes regardless of how suitable they are, so it is possible that none of these genes are suitable for estimating the contamination, or it may be that there are others not on the list more suitable.  It's purpose is to provide a heuristic to guide the user in deciding which genes to use.

Returning to our example, we see that the top three genes are LYZ, S100A9 and S100A8.  All three are reasonably bi-modal, but the circles representing individual cells indicate that there are still a number of cells where it is unclear whether the expression of the gene is due to contamination or the cell expressing it.  Furthermore, we see that the global contamination is rather high, both in an absolute sense (>10%) and relative to estimates from other genes.  It is often the case (although not always) that the most useful genes are those with the lowest average contamination estimate on this plot.

Proceeding down the list we come to IGKC.  Like the other three it has a clearly bimodal distribution and their are fewer cells that fall in the ambiguous region with a log ratio around 0.  It also produces a global estimate that is the lowest of all genes on this plot.  We also know that IGKC is a very specific gene expressed only in B-cells.  

This is already enough information to decide that IG genes are a good choice for estimating the contamination.  If we continue to look down the list we notice another gene, IGLC2, which is biologically related to IGKC and produces a very similar global estimate of the contamination.  This should give us even more confidence that IG genes are a good choice for this channel.

In practice, there are a few standard sets of genes, one or more of which is usually suitable for estimation of the contamination.  These are: immunoglobulin genes, haemoglobulin genes, MHC II genes, Mast cell genes TPSAB1/TPSB2, muscle specific genes.

# The lowest estimate that makes sense

The other use of the `inferNonExpressedGenes` and the `plotMarkerDistrubtion` functions is to help guide expectations of what contamination fraction makes sense.  If the estimate based on an individual gene is less than 10% (less than -1 on the `plotMarkerDistribution` plot), then any final estimates that exceed 10% should be treated with extreme suspicion.

If you are unable to get a contamination plot that is consistent with this expectation, it is preferable to force all cells to have a low contamination fraction via the `fixedContaminationValue` parameter of `interpolateCellContamination` than over-correct your data using dynamically calculated, inflated estimates of the contamination.
